README.md
# React Native Text Size

[![npm Version][npm-badge]][npm-url]
[![License][license-badge]][license-url]

Measure text accurately before laying it out and get font information from your App (Android and iOS).

In my country (México), software developers are poorly paid, so I have had to look for another job to earn a living and I cannot dedicate more time to maintaining this and other repositories that over the years have never generated any money for me. If anyone is interested in maintaining this repository, I'd be happy to transfer it to them, along with the associated npm package. |
:---: |
En mi país (México), los desarrolladores de software somos pésimamente pagados, por lo que he tenido que buscar otro trabajo para ganarme la vida y no puedo dedicar más tiempo a mantener éste y otros repositorios que a través de los años nunca me generaron dinero. Si a alguien le interesa dar mantenimiento a este repositorio, con gusto se lo transferiré, así como el paquete de npm asociado. |

There are two main functions: `flatHeights` to obtain the height of different blocks of text simultaneously, optimized for components such as [`<FlatList>`][0] or [`<RecyclerListView>`][1].

The other one is `measure`, which gets detailed information about one block of text:

- The width used by the text, with an option to calculate the real width of the largest line.
- The height of the text, with or without paddings.
- The number of lines.
- The width of the last line.
- Extended information of a given line.

The width and height are practically the same as those received from the `onLayout` event of a `<Text>` component with the same properties.

In both functions, the text to be measured is required, but the rest of the parameters are optional and work in the same way as with React Native:

- `fontFamily`
- `fontSize`
- `fontWeight`
- `fontStyle`
- `lineHeight`
- `numberOfLines`
- `fontVariant` (iOS)
- `includeFontPadding` (Android)
- `textBreakStrategy` (Android)
- `letterSpacing`
- `allowFontScaling`
- `width`: Constraint for automatic line-break based on text-break strategy.

In addition, the library includes functions to obtain information about the fonts visible to the App.

If it has helped you, please support my work with a star ⭐️ or [ko-fi][kofi-url].

## Installation

Mostly automatic installation from npm

```bash
yarn add react-native-text-size
react-native link react-native-text-size
```

Change the `compile` directive to `implementation` in the dependencies block of the android/app/build.gradle file.

**Requirements:**

- React Native v0.57 or later.
- Android API 16 or iOS 9.0 and above.

For versions prior to 0.56 of React Native, please use [react-native-text-size v2.1.1](https://www.npmjs.com/package/react-native-text-size/v/2.1.1)

See [Manual Installation][2] on the Wiki as an alternative if you have problems with automatic installation.

## API

- [`measure`](#measure)

- [`flatHeights` and `flatSizes`](#flatheights)

- [`specsForTextStyles`](#specsfortextstyles)

- [`fontFromSpecs`](#fontfromspecs)

- [`fontFamilyNames`](#fontfamilynames)

- [`fontNamesForFamilyName`](#fontnamesforfamilyname)

## measure

```ts
measure(options: TSMeasureParams): Promise<TSMeasureResult>
```

This function measures the text as RN does and its result is consistent\* with that of `Text`'s [onLayout](https://facebook.github.io/react-native/docs/text#onlayout) event. It takes a subset of the properties used by [`<Text>`][3] to describe the font and other options to use.

If you provide `width`, the measurement will apply automatic wrapping in addition to the explicit line breaks.

\* _There may be some inconsistencies in iOS, see this [Know Issue](#incorrent-height-ios) to know more._

**Note:**

Although this function is accurate and provides complete information, it can be heavy if the text is a lot, like the one that can be displayed in a FlatList. For these cases, it is better to use [`flatHeights`](#flatheights), which is optimized for batch processing.

### TSMeasureParams

Plain JS object with this properties (only `text` is required):

Property           | Type    | Default  | Notes
------------------ | ------  | -------- | ------
text               | string  | (none)   | This is the only required parameter and may include _emojis_ or be empty, but it **must not be** `null`.<br>If this is an empty string the resulting `width` will be zero.
fontFamily         | string  | OS dependent | The default is the same applied by React Native: Roboto in Android, San Francisco in iOS.<br>**Note:** Device manufacturer or custom ROM can change the default font.
fontWeight         | string  | 'normal' | On android, numeric ranges has no granularity and '500' to '900' becomes 'bold', but you can use a `fontFamily` of specific weight ("sans-serif-thin", "sans-serif-medium", etc).
fontSize           | number  | 14       | The default font size comes from RN.
fontStyle          | string  | 'normal' | One of "normal" or "italic".
lineHeight         | number  | (none)   | The line height of each line. Defaults to the font size.
numberOfLines      | number  | (none)   | Limit the number of lines the text can render on
fontVariant        | array   | (none)   | _iOS only_
ceilToClosestPixel | boolean | true     | _iOS only_. If true, we ceil the output to the closest pixel. This is React Native's default behavior, but can be disabled if you're trying to measure text in a native component that doesn't respect this.
allowFontScaling   | boolean | true     | To respect the user' setting of large fonts (i.e. use SP units).
letterSpacing      | number  | (none)   | Additional spacing between characters (aka `tracking`).<br>**Note:** In iOS a zero cancels automatic kerning.<br>_All iOS, Android with API 21+_
includeFontPadding | boolean | true     | Include additional top and bottom padding, to avoid clipping certain characters.<br>_Android only_
textBreakStrategy  | string  | 'highQuality' | One of 'simple', 'balanced', or 'highQuality'.<br>_Android only, with API 23+_
width              | number  | MAX_INT  | Restrict the width. The resulting height will vary depending on the automatic flow of the text.
usePreciseWidth    | boolean | false    | If `true`, the result will include an exact `width` and the `lastLineWidth` property.<br>You can see the effect of this flag in the [sample App][sample-app].
lineInfoForLine    | number  | (none)   | If `>=0`, the result will include a [lineInfo](#lineinfo) property with information for the required line number.

The [sample App][sample-app] shows interactively the effect of these parameters on the screen.

### TSMeasureResult

`measure` returns a Promise that resolves to a JS object with this properties:

Property      | Type   | Notes
------------- | ------ | ------
width         | number | Total used width. It may be less or equal to the `width` option.<br>On Android, this value may vary depending on the `usePreciseWidth` flag.
height        | number | Total height, including top and bottom padding if `includingFontPadding` was set (the default).
lastLineWidth | number | Width of the last line, without trailing blanks.<br>If `usePreciseWidth` is `false` (the default), this property is undefined.
lineCount     | number | Number of lines, taking into account hard and automatic line breaks.
lineInfo      | object | Line information.<br>If the `lineInfoForLine` option is not given, this property is undefined.

#### lineInfo

If the value of the `lineInfoForLine` is greater or equal than `lineCount`, this info is for the last line (i.e. `lineCount` - 1).

Property      | Type   | Notes
------------- | ------ | ------
line          | number | Line number of this info, base 0.<br>It can be less than the requested line number if `lineInfoForLine` is out of range.
start         | number | Text offset of the beginning of this line.
end           | number | Text offset after the last _visible_ character (so whitespace is not counted) on this line.
bottom        | number | The vertical position of the bottom of this line, including padding.
width         | number | Horizontal extent of this line, including leading margin indent, but excluding trailing whitespace.<br>Use `usePreciseWidth:true` to get an accurate value for this property.

In case of error, the promise is rejected with an extended Error object with one of the following error codes, as a literal string:

Code                 | Details
-------------------- | -------
E_MISSING_PARAMETERS | `measure` requires an object with the parameters, which was not provided.
E_MISSING_TEXT       | The text to measure is `null` or was not provided.
E_INVALID_FONT_SPEC  | The font specification is not valid. It is unlikely that this will happen on Android.
E_UNKNOWN_ERROR      | Well... who knows?

### Example

```jsx
//...
import rnTextSize, { TSFontSpecs } from 'react-native-text-size'

type Props = {}
type State = { width: number, height: number }

// On iOS 9+ will show 'San Francisco' and 'Roboto' on Android
const fontSpecs: TSFontSpecs = {
  fontFamily = undefined,
  fontSize = 24,
  fontStyle = 'italic',
  fontWeight = 'bold',
}
const text = 'I ❤️ rnTextSize'

class Test extends Component<Props, State> {
  state = {
    width: 0,
    height: 0,
  }

  async componentDidMount() {
    const width = Dimensions.get('window').width * 0.8
    const size = await rnTextSize.measure({
      text,             // text to measure, can include symbols
      width,            // max-width of the "virtual" container
      ...fontSpecs,     // RN font specification
    })
    this.setState({
      width: size.width,
      height: size.height
    })
  }

  // The result is reversible
  render() {
    const { width, height } = this.state
    return (
      <View style={{ padding: 12 }}>
        <Text style={{ width, height, ...fontSpecs }}>
          {text}
        </Text>
      </View>
    )
  }
}
```

## flatHeights

```ts
flatHeights(options: TSHeightsParams): Promise<number[]>
flatSizes(options: TSHeightsParams): Promise<TSFlatSizes>
```

Calculate the height (or sizes) of each of the strings in an array.

This is an alternative to `measure` designed for cases in which you have to calculate the height of numerous text blocks with common characteristics (width, font, etc), a typical use case with `<FlatList>` or `<RecyclerListView>` components.

The measurement uses the same algorithm as `measure` but it returns only the height of each block and, by avoiding multiple steps through the bridge, it is faster... _much faster_ on Android!

I did tests on 5,000 random text blocks and these were the results (ms):

&nbsp;  | `measure` | `flatHeights`
------- | --------: | ----------:
Android | 49,624    | 1,091
iOS     |  1,949    |   732

Note that `flatSizes` is somewhat slower than `flatHeights` on Android since it needs to iterate through lines to find the maximum line length.

In the future I will prepare an example of its use with FlatList and multiple styles on the same card.

### TSHeightsParams

This is an object similar to the one you pass to `measure`, but the `text` option is an array of strings and the `usePreciseWidth` and `lineInfoForLine` options are ignored.

Property            | Type     | Default
------------------- | -------- | --------
text                | string[] | (none)
width               | number   | Infinity
fontFamily          | string   | OS dependent
fontWeight          | string   | 'normal'
fontSize            | number   | 14
fontStyle           | string   | 'normal'
fontVariant         | array    | (none)
allowFontScaling    | boolean  | true
letterSpacing       | number   | (none)
includeFontPadding  | boolean  | true
textBreakStrategy   | string   | 'highQuality'
numberOfLines       | number   | (none)

The result is a Promise that resolves to an array with the height of each block (in _SP_), in the same order in which the blocks were received.

Unlike measure, `null` elements returns 0 without generating error, and empty strings returns the same height that RN assigns to empty `<Text>` components (the difference of the result between `null` and empty is intentional).

### Example

```jsx
//...
import rnTextSize, { TSFontSpecs } from 'react-native-text-size'

type Props = { texts: string[] }
type State = { heights: number[] }

// On iOS 9+ will show 'San Francisco' and 'Roboto' on Android
const fontSpecs: TSFontSpecs = {
  fontFamily = undefined,
  fontSize = 24,
  fontStyle = 'italic',
  fontWeight = 'bold',
}
const texts = ['I ❤️ rnTextSize', 'I ❤️ rnTextSize using flatHeights', 'Thx for flatHeights']

class Test extends Component<Props, State> {
  state = {
    heights: [],
  }

  async componentDidMount() {
    const { texts } = this.props
    const width = Dimensions.get('window').width * 0.8
    const heights = await rnTextSize.flatHeights({
      text: texts,      // array of texts to measure, can include symbols
      width,            // max-width of the "virtual" container
      ...fontSpecs,     // RN font specification
    })
    this.setState({
      heights
    })
  }

  render() {
    const { texts } = this.props
    const { heights } = this.state
    
    return (
      <View style={{ padding: 12 }}>
        {texts.map(
          (text, index) => (
            <Text style={{ height: heights[index], ...fontSpecs }}>
              {text}
            </Text>
          )
        )}
      </View>
    )
  }
}
```

## specsForTextStyles

```ts
specsForTextStyles(): Promise<{ [key: string]: TSFontForStyle }>
```

Get system font information for the running OS.

This is a wrapper for the iOS [`UIFont.preferredFontForTextStyle`][4] method and the current Android [Material Design Type Scale][5] styles.

The result is a Promise that resolves to a JS object whose keys depend on the OS, but its values are in turn objects fully compatible with those used in the RN styles, so it can be used to stylize `<Text>` or `<TextInput>` components:

### TSFontForStyle

Property      | Type         | Notes
------------- | ------------ |------
fontFamily    | string       | System family name or font face.
fontSize      | number       | Font size in _SP_ (unscaled).
fontStyle     | TSFontStyle  | Only if 'italic', undefined if the style is 'normal'.
fontWeight    | TSFontWeight | Only if 'bold', undefined if the weight is 'normal'.
fontVariant   | TSFontVariant[] or null | _iOS only_. Currently, no style includes this property.
letterSpacing | number       | Omitted if running on Android with RN lower than 0.55

To know the key names, please see [Keys from specsForTextStyles][6] in the Wiki.

I have not tried to normalize the keys of the result because, with the exception of two or three, they have a different interpretation in each OS, but you can use them to create custom styles according to your needs.

## fontFromSpecs

```ts
fontFromSpecs(specs: TSFontSpecs): Promise<TSFontInfo>
```

Returns the characteristics of the font obtained from the given specifications.

### TSFontSpecs

This parameter is a subset of [`TSMeasureParams`](#tsmeasureparams), so the details are omitted here.

Property      | Type     | Default
------------- | -------- | -------
fontFamily    | string   | iOS: 'San Francisco', Android: 'Roboto'
fontWeight    | string   | 'normal'
fontSize      | number   | 14
fontStyle     | string   | 'normal'
fontVariant   | string[] | (none)
letterSpacing | number   | 0

`fontFromSpecs` uses an implicit `allowsFontScaling:true` and, since this is not a measuring function, `includeFontPadding` has no meaning.

### TSFontInfo

The result is a Promise that resolves to a JS object with info for the given font and size, units in [_SP_][7] in Android or points in iOS, using floating point numbers where applicable\*.

Property    | Type     | Details
----------- | -------- | --------
fontFamily  | string   | In Android it is the same string passed as parameter.
fontName    | string   |_iOS only_, always `undefined` in Android.
fontSize    | number   | It may be different from the given parameter if the parameter includes decimals.
fontStyle   | string   | 'normal' or 'italic'.
fontWeight  | string   | 'normal' or 'bold', on iOS it can go from '100' to '900'.
fontVariant | string[] | _iOS only_, always `undefined` in Android.
ascender    | number   | The recommended distance above the baseline for singled spaced text.
descender   | number   | The recommended distance below the baseline for singled spaced text.
capHeight   | number   | _iOS only_ Height of capital characters.
xHeight     | number   | _iOS only_ Height of lowercase "x".
top         | number   | _Android only_. Maximum distance above the baseline for the tallest glyph in the font.
bottom      | number   | _Android only_. Maximum distance below the baseline for the lowest glyph in the font.
leading     | number   | The recommended additional space to add between lines of text.
lineHeight  | number   | The recommended line height. It should be greater if text contain Unicode symbols, such as emojis.

CHANGELOG.md
# Changelog

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## \[Unreleased]

### Added

- Update README.md with example for flatHeights - Thanks to @donni106
- Create new flatSizes function that also gets the widths
- Add support for `numberOfLines` and `maxWidth/maxHeight` dimensions.

### Changed

- The dependency on android.support.annotations is replaced by javax.annotation

### Fixed

- fix s.source_files regex to match RNTextSize source files: RNTextSize - Thanks to @abegehr

## \[3.0.0] - 2019-01-10

### Added

- Support `textTransform:uppercase` for specsForTextStyles in Android.

### Changed

- peerDependency on react-native to >=57.0
- Update gradle files to be consistent with RN 0.57
- Update License year.

### Removed

- Unused 'example' directory.

## \[2.1.1] - 2019-01-10

### Added

- Markdown lint rules.
- `lineInfoForLine` option, to get information for a given line.

### Changed

- Update Readme.
- Limit peerDependency on RN to v0.56, rnTextSize v3 will support RN>=57

### Fixed

- Linting errors in the markdown of the Changelog and Readme files.

## \[2.0.4] - 2018-09-14

### Changed

- Patch to v2.0.3 published from wrong branch

## \[2.0.3] - 2018-09-14

### Changed

- Updated README

### Fixed

- PR #6 Fix podspec sysntax. Thanks to @Jyrno42

## \[2.0.2] - 2018-08-31

### Changed

- iOS: Give `flatHeights` better performace by avoiding multiple creation of NSTextStorage.
- Android: Now Studio can check versions in android/build.gradle

### Fixed

- Closes #5: Inconsistency in flatHeights between Android and iOS.
- Some error in the README and include note about unlink the previous version.

## \[2.0.1] - 2018-08-22

### Fixed

- Silly typo error in last minute edition.

## \[2.0.0] - 2018-08-22

Bump v2.0.0 :tada:

### Changed

- Code cleanup, minor enhancements.

### Removed

- Removed the `FontSize` constants.

## \[2.0.0-beta.6] - 2018-08-21

### Added

- New function `flatHeights` to calculate the height of multiple strings at once, much faster than `measure`.
- Revised Readme. Now it's clearer, part of its content moved to the Wiki.
- Docummented the iOS-only properties `capHeight` and `xHeight` from the `fontFromSpecs` result.
- New flag `usePreciseWidth` (default `false`) request the most accurate calculation of the width (Android) and the value of `lastWidth` (both), but its is a bit slower.

### Changed

- iOS: The `specsForTextStyles` function returns fontSize amd letterSpacing with unscaled values, to allow its use with `allowFontScaling`.

### Fixed

- Android: `Arguments.fromList` does not exists in RN 0.52, replaced with `Arguments.fromArray`
- iOS: Fix errors in the sample App that prevented it from running in iOS.

### Removed

- To avoid interfering with this changelog, the sample application was moved to its own repository.

## \[2.0.0-beta.4] - Unpublished

**WARNING:**

_The package id has changed, you must uninstall the previous version before using this one._

### Added

- Note on the README of the sample App, to test it with a different version of RN.
- Adds flow typings (almost) working with Flow 0.61.x
- Enhancements to the sample App with RN 0.56 and Flow
- New functions `specsForTextStyles`, `fontFromSpecs`, and `fontFamilyNames`

### Changed

- Complete rewrite of the sample app.
- Simplify anroid/build.gradle and adjust the configuration.
- `allowFontScaling` is handled in the native side, index.js was simplified a lot.
- Migration of example to RN 0.52 and Typescript, enhanced features.
- Minimum supported versions: RN 0.52.x, Android SDK 16, iOS 9.0
- New fformat of the CHANGELOG to follow the _Keep a Changelog_ recommendations.
- Package ID now is `com.github.amarcruz.rntextsize` for Android, and `RNTextSize` for iOS.
- Fixes to typings.

### Removed

- Drop support for iOS 8 and RN below 0.52

### Fixed

- Tested on iOS 9 & 11 simulators, Android devices 5.1 & 7, simulator 4.4 & 11, all with RN 0.52.0 and 0.56.0
- Improved measurement on both platforms, accuracy is more than 5 decimals.
- The iOS API and behavior of `measure` now is the same as Android.
- Makes the example works.

## \[1.0.0-beta.5] - Unpublished

- Preliminary implementation for iOS (thanks to @xuhom)
- The license is changed to BSD 2-Clause as [react-native-measure-text](https://github.com/airamrguez/react-native-measure-text), on which this library is based.

## \[1.0.0-beta.4] - Unpublished

### Changed

- Now the SDK versions are compatible with global rootProject.ext for flexible configuration.

## [1.0.0-beta.3] - 2018-02-17

Published to npm, tested with Android SDK 21.

### Changed

- Updates compileSdkVersion, buildToolsVersion and targetSdkVersion.
- Gradle plugin to 2.3.3

## \[1.0.0-beta.2] - Unpublished

### Changed

- Using ES6 for index.js

### Fixed

- Correction to the default export in the index

## \[1.0.0-alpha.10] - Unpublished

### Changed

- compileSdkVersion 25, buildToolsVersion 25.0.3, targetSdkVersion 25

## \[1.0.0-alpha.9] - Unpublished

### Changed

- Revision to README
- targetSdkVersion from 22 to 23.

### Fixed

- Fix error in android path
- Android package id is 'io.amarcruz.rnmeasuretext'

## \[1.0.0-alpha.8] - Unpublished

### Added

- Suppport for `includeFontPadding`

### Changed

- Using gradle 2.2.3

## \[1.0.0-alpha.7] - Unpublished

### Added

- Returned info includes `lineCount` with the number of lines and `lastLineWidth` with the last line width (Android).
- Uses scaled `fontSize` (with `allowFontScaling`, automatic).

### Changed

- Makes `width` property optional in parameters (Android).
- Minor fixes, now working.
- First commit, Android only.


example/App.js
/* @flow */
/* eslint max-len:0 */
import React from 'react'
import {
  Dimensions,
  Picker,
  Platform,
  SafeAreaView,
  StatusBar,
  ScrollView,
  StyleSheet,
  Switch,
  //TabBarIOS,
  ToolbarAndroid,
  Text,
  TextInput,
  View,
} from 'react-native'
import TextSize from 'react-native-text-size'

import type { ViewLayoutEvent } from 'ViewPropTypes'
import type { TSMeasureResult, TSFontSpecs, TSFontVariant, TSTextBreakStrategy } from 'react-native-text-size'

type PickerItemList = Array<string | void>
type Props = {}
type State = {
  info?: TSMeasureResult | null,
  specs: TSFontSpecs,
  text: string,
  width: number,
  allowFontScaling: boolean,
  fonts?: string[] | void,
  layout?: { width: number, height: number }
}

const IOS = Platform.OS === 'ios' && Platform.Version || undefined
const ANDROID = Platform.OS === 'android' && Platform.Version || undefined

const winDims = Dimensions.get('window')
const TEXT_TOP = 0
const TEXT_LEFT = 16
const TEXT_WIDTH = 274

const TEST_FONT: TSFontSpecs = {
  fontFamily: 'GreatVibes-Regular',
  fontSize: 28,
  fontStyle: undefined,
  fontWeight: undefined,
  fontVariant: undefined,
  includeFontPadding: true,
  letterSpacing: undefined,
  textBreakStrategy: undefined,
}
const TEXT_STR = 'This is a first string\n' +
'The second string is slightly bigger ƒƒ \n' +
'Bacon ⌛ ⌨ ☄ 🤘 ipsum dolor 12345 amet 67890 capicola filet mignon flank venison ball tip pancetta cupim tenderloin bacon beef shank.'

const reactNativeVersion = (): { major: number, minor: number, patch: number } => {
  try {
    return require('./node_modules/react-native/Libraries/Core/ReactNativeVersion').version
  } catch (_) {
    return { major: 0, minor: 0, patch: 0 } // bellow 0.49.x
  }
}

// 5 decimals max
const formatNumber = (n: number) => {
  return n.toFixed(6).replace(/\.?0+$/, '')
}

export default class MeasureApp extends React.Component<Props, State> {

  constructor (props: Props) {
    super(props)

    this.state = {
      specs: TEST_FONT,
      text: TEXT_STR,
      width: TEXT_WIDTH,
      allowFontScaling: true,
    }

    if (ANDROID) {
      StatusBar.setBackgroundColor('#002984')
      StatusBar.setTranslucent(true)
    }

    TextSize.fontFromSpecs({
      ...TEST_FONT,
    }).then((info) => {
      console.log('Test font info:', info)
    }).catch(console.warn)

    const rn = reactNativeVersion()
    console.log(`Dimensions scale: ${winDims.scale}, fontScale: ${winDims.fontScale}`)
    console.log(`ReactNative ${rn.major}.${rn.minor}.${rn.patch}`)
    // $FlowSucks - can't convert undefined to string???
    console.log(`${IOS ? 'iOS' : 'Android SDK'} ${IOS || ANDROID}`)

    TextSize.specsForTextStyles()
      .then((specs) => { console.log('specsForTextStyles: ', specs) })
      .catch(console.error)
    //console.log('TextSize.FontSize:', TextSize.FontSize)
    //TextSize.fontForTextStyle('body').then(info => {
    //  console.log(`Font info for body:`, info)
    //})
    //TextSize.fontForFontType('default').then(info => {
    //  console.log(`Font info of "default":`, info)
    //})
  }

  componentDidMount () {
    TextSize.measure({
      text: this.state.text,
      width: this.state.width,
      allowFontScaling: this.state.allowFontScaling,
      ...this.state.specs,
    }).then((info) => {
      console.log('TextSize full info:', info)
      this.displayResult(info)
      this.setState({ info })
    }).catch((err) => {
      console.warn('Error in measure:', err)
    })
    TextSize.fontFamilyNames().then((fonts) => {
      this.setState({ fonts })
    })
  }

  displayResult (info: TSMeasureResult) {
    console.log(`TextSize info - height: ${info.height}, width: ${
      info.width}, lastLineWidth: ${info.lastLineWidth}, lineCount: ${info.lineCount}`)
  }

  doMeasure (prop: $Shape<TSFontSpecs>) {
    const { specs, text, width, allowFontScaling } = this.state
    const specsParams = {
      ...specs, text, width, allowFontScaling, ...prop,
    }
    TextSize.measure(specsParams).then((info) => {
      this.displayResult(info)
      this.setState((state) => {
        return {
          specs: { ...state.specs, ...prop },
          info,
        }
      })
    }).catch(console.error)
  }

  setFontFamily = (fontFamily: string) => {
    this.doMeasure({ fontFamily })
  }
  setFontStyle = (fontStyle: $PropertyType<TSFontSpecs, 'fontStyle'>) => {
    this.doMeasure({ fontStyle: fontStyle || undefined })
  }
  setFontWeight = (fontWeight: $PropertyType<TSFontSpecs, 'fontWeight'>) => {
    this.doMeasure({ fontWeight: fontWeight || undefined })
  }
  setFontSize = (ev: any) => {
    const fs = parseFloat(ev.nativeEvent.text)
    this.doMeasure({ fontSize: !isNaN(fs) && fs > 0 ? fs : undefined })
  }
  setLetterSpacing = (ev: any) => {
    const fs = parseFloat(ev.nativeEvent.text)
    this.doMeasure({ letterSpacing: fs || undefined })
  }
  setTextBreakStrategy = (textBreakStrategy: TSTextBreakStrategy) => {
    this.doMeasure({ textBreakStrategy })
  }
  setIncludeFontPadding = (includeFontPadding: boolean) => {
    this.doMeasure({ includeFontPadding })
  }
  setFontVariant = (variant: TSFontVariant) => {
    this.doMeasure({ fontVariant: variant ? [variant] : undefined })
  }

  setAllowFontScaling = (allowFontScaling: boolean) => {
    const { specs, text, width } = this.state
    const specsParams = {
      ...specs, text, width, allowFontScaling,
    }
    TextSize.measure(specsParams).then((info) => {
      this.displayResult(info)
      this.setState({ allowFontScaling })
    }).catch(console.error)
  }

  onLayout = (e: ViewLayoutEvent) => {
    const info = e.nativeEvent.layout
    this.setState({ layout: { width: info.width, height: info.height } })
    console.log(`onLayout info - height: ${info.height}, width: ${info.width}`)
  }

  renderPickerItems (items: PickerItemList) {
    // $FlowSucks ...come on
    return items.map((str) => {
      const name = str === undefined ? '(undefined)' : str
      return  <Picker key={`key-${name}`} label={name} value={str || ''} />
    })
  }

  render () {
    const { info, specs, text, allowFontScaling, fonts, layout } = this.state

    let sizes, infoStat, posStyle
    if (info) {
      sizes = {
        height: info.height,
        width: info.width,
        minHeight: info.height,
        minWidth: info.width,
      }
      posStyle = { left: info.lastLineWidth, top: TEXT_TOP + info.height }
      infoStat = `TextSize height: ${formatNumber(info.height)}, width: ${formatNumber(
        info.width)}\n  lastLineWidth: ${formatNumber(info.lastLineWidth)} lines: ${info.lineCount}`
    } else {
      sizes = { width: -1 }
      posStyle = { left: 10, top: 50 }
      infoStat = 'waiting for text-size...'
    }

    const layoutStat = layout ? `onLayout height: ${
      formatNumber(layout.height)}, width: ${formatNumber(layout.width)} ` : ' '

    return (
      <SafeAreaView style={styles.container}>

        <ToolbarAndroid title="TextSize Tester" titleColor="#fff" style={{
          height: 56,
          backgroundColor: '#3f51b5',
          elevation: 6,
        }} />

        <ScrollView style={{
          paddingLeft: TEXT_LEFT,
          paddingRight: TEXT_LEFT - 8,
          marginTop: 4,
        }}>

          <View style={styles.row}>
            <Text style={styles.prompt}>Font:</Text>
            <Picker
              mode="dialog"
              prompt="Select fontFamily"
              style={styles.pickerBox}
              selectedValue={specs.fontFamily}
              onValueChange={this.setFontFamily}
              key="fontFamily"
            >
              {fonts && this.renderPickerItems([undefined, ...fonts])}
            </Picker>
          </View>
          <View style={styles.row}>
            <Text styles={styles.prompt}>fontStyle:</Text>
            <Picker
              style={styles.pickerBox}
              selectedValue={specs.fontStyle}
              onValueChange={this.setFontStyle}
            >
              {this.renderPickerItems([undefined, 'normal', 'italic'])}
            </Picker>
          </View>
          {IOS &&
            <View style={styles.row}>
              <Text styles={styles.prompt}>fontVariant:</Text>
              <Picker
                style={styles.pickerBox}
                selectedValue={specs.fontVariant && specs.fontVariant[0] || ''}
                onValueChange={this.setFontVariant}>
                {this.renderPickerItems([undefined,
                  'small-caps', 'oldstyle-nums', 'lining-nums', 'tabular-nums', 'proportional-nums'])}
              </Picker>
            </View>
          }
          <View style={styles.row}>
            <Text styles={styles.prompt}>fontWeight:</Text>
            <Picker
              style={styles.pickerBox}
              selectedValue={specs.fontWeight}
              onValueChange={this.setFontWeight}>
              {this.renderPickerItems([undefined, 'normal', 'bold'])}
            </Picker>
          </View>
          <View style={styles.row}>
            <Text styles={styles.prompt}>fontSize:</Text>
            <TextInput
              ref="fontSizeInput"
              style={styles.numeric}
              autoCapitalize="none"
              keyboardType="numeric"
              placeholder="enter size"
              defaultValue={specs.fontSize && specs.fontSize > 0 ? String(specs.fontSize) : ''}
              key="fontSize"
              onEndEditing={this.setFontSize}
            />
          </View>
          <View style={styles.row}>
            <Text styles={styles.prompt}>letterSpacing:</Text>
            <TextInput
              ref="letterSpacingInput"
              style={styles.numeric}
              autoCapitalize="none"
              keyboardType="number-pad"
              placeholder="spacing"
              defaultValue={specs.letterSpacing ? String(specs.letterSpacing) : ''}
              onEndEditing={this.setLetterSpacing}
            />
          </View>
          {ANDROID &&
            <View style={styles.row}>
              <Text styles={styles.prompt}>textBreakStrategy:</Text>
              <Picker
                style={styles.pickerBox}
                selectedValue={specs.textBreakStrategy}
                onValueChange={this.setTextBreakStrategy}
              >
                {this.renderPickerItems([undefined, 'highQuality', 'balanced', 'simple'])}
              </Picker>
            </View>
          }
          <View style={styles.row}>
            <Text styles={styles.prompt}>allowFontScaling:</Text>
            <Switch
              style={styles.switchBox}
              value={allowFontScaling}
              onValueChange={this.setAllowFontScaling}
            />
          </View>

          {ANDROID &&
            <View style={styles.row}>
              <Text styles={styles.prompt}>includeFontPadding:</Text>
              <Switch
                style={styles.switchBox}
                value={specs.includeFontPadding}
                onValueChange={(includeFontPadding) => this.doMeasure({ includeFontPadding })}
              />
            </View>
          }
          <View style={styles.lastRow}>
            <Text style={styles.statusText}>{layoutStat}</Text>
            <Text style={styles.statusText}>{infoStat}</Text>
          </View>

          {/*
            Graphical Output
          */}
          <View>
            {sizes.width > 0 && <View style={[styles.result, sizes]} />}

            <Text
              allowFontScaling={allowFontScaling}
              style={[styles.sample, (specs: any)]}
              onLayout={this.onLayout}>
              {text}
            </Text>

            <View style={[styles.lastLineWidthMark, posStyle]} />
          </View>

        </ScrollView>
      </SafeAreaView>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    //...StyleSheet.absoluteFillObject,
    //paddingTop: (StatusBar.currentHeight | 0) + 8,
    ...Platform.select({
      android: {
        marginTop: StatusBar.currentHeight | 0,
      },
    }),
  },
  row: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    minHeight: 40,
    borderBottomWidth: 1,
    borderBottomColor: '#ccc',
  },
  lastRow: {
    justifyContent: 'space-between',
    minHeight: 44,
    paddingTop: 8,
    paddingBottom: 12,
  },
  prompt: {
    //height: '100%',
    textAlignVertical: 'center',
  },
  statusText: {
    fontFamily: IOS ? 'Courier' : 'monospace',
    fontSize: 12,
  },
  sample: {
    top: TEXT_TOP,
    left: 0,
    width: TEXT_WIDTH,
    maxWidth: TEXT_WIDTH,
    marginBottom: 40,
    backgroundColor: 'rgba(255,0,0,0.25)',
  },
  result: {
    position: 'absolute',
    top: TEXT_TOP,
    left: 0,
    borderColor: 'black',
    borderWidth: 1,
  },
  lastLineWidthMark: {
    position: 'absolute',
    width: 2,
    borderLeftWidth: 1,
    borderLeftColor: '#0000cc',
    height: 24,
  },
  pickerBox: {
    flexGrow: 1,
    marginLeft: 4,
  },
  numeric: {
    minWidth: 128,
    textAlign: 'right',
    marginRight: 12,
    fontFamily: IOS ? 'Courier' : 'monospace',
    fontWeight: 'bold',
  },
  switchBox: {
    marginRight: 8,
  },
})
